CiliumNetworkPolicy validation

romanspb80 opened the pull request:
This patch adds validation for CiliumNetworkPolicy with CEL

Fixes: #23152

kaworu commented on install/kubernetes/cilium/values.yaml:
I don't think removing `k8sClientRateLimit` was intended, also please report your changes to `install/kubernetes/cilium/values.yaml.tmpl` and then run:

```
make -C install/kubernetes cilium/values.yaml
```

to update this file.

joamaki replied:
`k8sClientRateLimit` was just introduced, so rebase and then regenerate `values.yaml`.

kaworu reviewed (changes requested):
Thanks for the PR @romanspb80, please see my comment on `value.yaml`. Also pulling @cilium/sig-policy for the review.

gandro reviewed (approved):
In terms of Helm structure, this looks good to me.

I'll defer the review of the actual validation logic to other subteams though.

giorio94 commented on install/kubernetes/cilium/templates/cilium-agent/validatingadmissionpolicy.yaml:
I'd suggest to move these two files to a dedicated directory, renaming the files to reflect the name of the policy.

giorio94 commented on install/kubernetes/cilium/templates/cilium-agent/validatingadmissionpolicybinding.yaml:
When trying to apply this policy, an error is returned:

```
The ValidatingAdmissionPolicyBinding "cilium-validate-network-policies-binding" is invalid: spec.validationActions: Required value: at least one validation action is required
```

giorio94 reviewed (changes requested):
I made a try applying this validation policy, but it appears to reject even empty policies:

```
cat <<EOF | kubectl apply -f - 
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector: {}
EOF
```

Returns: 

```
The ciliumnetworkpolicies "test" is invalid: : ValidatingAdmissionPolicy 'cilium-validate-network-policies' with binding 'cilium-validate-network-policies-binding' denied request: L7 rules can only apply to TCP (not UDP) except for egress DNS rules
```

My understanding is that an operation is rejected if the expression evaluates to false, and not when it evaluates to true.

I wonder how we could have tests to ensure that this (and other policies which we may introduce in the future) actually work as expected, and to prevent possible future regressions. 

I'd be also personally hesitant in saying that this PR fixes https://github.com/cilium/cilium/issues/23152. While being a step forward for sure, I feel that there may be other combinations of settings which would benefit from a validation, as invalid.

aditighag commented on install/kubernetes/cilium/templates/cilium-agent/validatingadmissionpolicy.yaml:
```diff
@@ -0,0 +1,56 @@
+        operations:  ["CREATE", "UPDATE"]
+        resources:   ["ciliumnetworkpolicies", "ciliumclusterwidenetworkpolicies"]
+  validations:
+    - expression:
+        (
+          has(object.spec.ingress)
+          && object.spec.ingress.exists(x, has(x.toPorts)
+            && x.toPorts.exists(x,
+              (
+                has(x.ports) && (
+                  x.ports.exists(x, x.protocol == 'TCP')
+                  && (
+                       has(x.rules) && (
+                         has(x.rules.http) || has(x.rules.kafka)
+                       )
+                  )
+                ) && (
+                       has(x.rules) && !has(x.rules.dns)
+                )
+              )
+            )
+          )
+        ) ||
+        (
+          has(object.spec.egress)
+          && object.spec.egress.exists(x, has(x.toPorts)
+            && x.toPorts.exists(x,
+              (
+                has(x.ports) && (
+                  x.ports.exists(x, x.protocol == 'TCP')
+                  && (
+                       has(x.rules) && (
+                         has(x.rules.http) || has(x.rules.kafka)
```
Can you add a reference link to the CNP spec here so that readers know where these rules come from?

aditighag commented on install/kubernetes/cilium/templates/cilium-agent/validatingadmissionpolicy.yaml:
```diff
@@ -0,0 +1,56 @@
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "cilium-validate-network-policies"
```
```suggestion
  name: "validate-cilium-network-policies"
```

aditighag reviewed (changes requested):
Very cool! 

>This patch adds validation for CiliumNetworkPolicy with CEL

Please spell out CEL in the description, and add an official k8s documentation link. It'll be great add a reference link to the CNP spec where the rules are specified.

Can you also add an example in the PR description showing how the validation fails an incorrect CNP?

romanspb80 committed:
CiliumNetworkPolicy validation

This patch adds validation for CiliumNetworkPolicy with CEL
CiliumNetworkPolicy: https://docs.cilium.io/en/stable/security/policy/language/#layer-7-examples
Validating Admission Policy 9with CEL): https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/

For example if field "protocol" is set "UDP" or "ANY":
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "rule1"
spec:
  description: "Allow HTTP GET /public from env=prod to app=service"
  endpointSelector:
    matchLabels:
      app: service
  ingress:
  - fromEndpoints:
    - matchLabels:
        env: prod
    toPorts:
    - ports:
      - port: "80"
        protocol: UDP
      rules:
        http:
        - method: "GET"
          path: "/public"

then "validate-cilium-network-policies" deny this request with the warning:
"L7 rules can only apply to TCP (not UDP) except for egress DNS rules"

Fixes: #23152

Signed-off-by: Roman Ptitcyn <romanspb@yahoo.com>

nathanjsweet commented on install/kubernetes/cilium/templates/cilium-agent/validatingadmissionpolicy.yaml:
```diff
@@ -0,0 +1,59 @@
+    - expression:
+        (
+          has(object.spec.ingress)
+          && object.spec.ingress.exists(x, has(x.toPorts)
```
Members of `ingress` do not need `toPorts`. My proposed change checks if `toPorts` does not exist and then proceeds to validate `toPorts` if it does.
```suggestion
          && object.spec.ingress.exists(x, !has(x.toPorts)
```

nathanjsweet commented on install/kubernetes/cilium/templates/cilium-agent/validatingadmissionpolicy.yaml:
```diff
@@ -0,0 +1,59 @@
+        (
+          has(object.spec.ingress)
+          && object.spec.ingress.exists(x, has(x.toPorts)
+            && x.toPorts.exists(x,
```
With the above changes this converts to an "or".
```suggestion
            x.toPorts.exists(x,
```

nathanjsweet reviewed (changes requested):
Your expression is not structured to cover both `egress` and `ingress` if they are both present (and if they are both present and one of them is invalid, then this validation logic will miss the invalid object), the `toPorts` validation incorrectly assumes that `toPorts` must be present but there are other membership possibilities to consider, and it doesn't validate `endpointSelectors`.

The structure of the expression should look something like this:
```cel
has(object.spec.endpointSelectors) &&
// validate endpointSelectors
(... ) &&
 // validate ingress object
(!has(object.spec.ingress) ||
    ...
) &&
// validate egress object
(!has(object.spec.egress) ||
    ...
)
```

I'm converting this PR to a draft as it needs more work.

Feel free to reach out on slack for more detailed feedback.
